src/Halogen/Query/EventSource.purs

```diff
@@ -0,0 +1,72 @@
+module Halogen.Query.EventSource
+  ( EventSourceE()
+  , EventSource()
+  , makeEventSource
+  , runEventSource
+  , interpretEventSource
+  , catEventSource
+  , hoistEventSource
+  , contrahoistEventSource
+  ) where
+
+import Prelude
+
+import Control.Coroutine.Stalling as SCR
+import Control.Monad.Aff (Aff())
+import Control.Monad.Free.Trans (hoistFreeT, interpret)
+
+import Data.Bifunctor (lmap)
+import Data.ExistsRE (ExistsRE(), mkExistsRE, runExistsRE)
+import Data.NaturalTransformation (Natural())
+import Data.Functor.Coproduct (Coproduct(), coproduct)
+import Data.Const (Const())
+import Data.Maybe (Maybe(..))
+
+import Unsafe.Coerce as U
+
+newtype EventSourceE f eff = EventSourceE (SCR.StallingProducer (f Unit) (Aff eff) Unit)
+newtype EventSource f (g :: * -> *) = EventSource (ExistsRE (EventSourceE f))
+
+makeEventSource :: forall f eff. SCR.StallingProducer (f Unit) (Aff eff) Unit -> EventSource f (Aff eff)
+makeEventSource = EventSource <<< mkExistsRE <<< EventSourceE
+
+runEventSource :: forall f eff. EventSource f (Aff eff) -> SCR.StallingProducer (f Unit) (Aff eff) Unit
+runEventSource (EventSource xese) = runExistsRE (\(EventSourceE p) -> hoistFreeT coerceEffects p) xese
+  where
+  -- This coercion is safe because we can only construct an `EventSource`
+  -- with an `Aff`, and we can only transform `g` with natural tranformations,
+  -- so for any `Aff` we end up with the effect row must subsume the original.
```

Comment:
If there is a possibility that the correctness of this approach only depends on a weaker principle than the one I laid out, maybe it can be salvaged...

But I am feeling a little bit pessimistic about it.